text/0034-ternary-hash.md

```diff
@@ -0,0 +1,415 @@
+verified by nodes in the network. Conversely, a node needs to be able to verify other nodes messages. Among other
+operations, this is accomplished by checking transaction signatures (from which addresses are calculated), transaction
+hashes, and hashes of collections of transactions (so-called bundles).
+
+The two hash functions currently used are both sponge constructions: `CurlP`, which is specified entirely in balanced
+ternary, and `Kerl`, which first converts ternary input to a binary representation, applies `keccak-384` to it, and then
+converts its binary output back to ternary. For `CurlP` specifically, its variants `CurlP27` and `CurlP81` are used.
+
+# Detailed design
+
+## Hash
+
+This RFC defines a ternary type `Hash`. The exact definition is implementation detail but an example definition could
+simply be the following.
+
+```rust
+struct Hash([i8; 243]);
+```
+
+## Sponges
+
+`CurlP` and `Kerl` are cryptographic sponge constructions. They are equipped with a memory state and a function that
+replaces the state memory using some input string (which can be the state memory itself). A portion of the memory state
+is then the output. In the sponge metaphor, the process of replacing the memory state by an input string is said to
+`absorb` the input, while the process of producing an output is said to `squeeze` out an output.
+
+The hash functions are expected to be used like this:
+
+```rust
+// Create a CurlP instance with 81 rounds.
+// This is equivalent to calling `CurlP::new(81)`.
+let mut curlp = CurlP81::new();
+
+// Assume we have some input that is a Trits slice as defined by the bee_ternary crate, and an output buffer `TritBuf`:
+let transaction: &Trits<T1B1>;
+let mut tx_hash: TritBuf<T1B1Buf>;
+
+// Absorb the transaction.
+curlp.absorb(&Trits::from_i8_unchecked(&transaction));
+// And squeeze into the `tx_hash` buffer.
+curlp.squeeze_into(&mut tx_hash);
+```
+```rust
+// Create a Kerl instance. Kerl does not come with a configurable number of rounds.
+let mut kerl = Kerl::new();
+
+// `Kerl::digest` is a function that combines `Kerl::absorb` and `Kerl::squeeze`.
+// `Kerl::digest_into` combines `Kerl::absorb` with `Kerl::squeeze_into`. `CurlP` provides the same methods.
+let tx_hash = kerl.digest(&transaction);
+```
+
+The main proposal of this RFC are the `Sponge` trait and the `CurlP` and `Kerl` types that are implemented in terms of
+it. This RFC relies on the the presence of the types `TritBuf` and `Trits`, which are assumed to be owning and borrowing
+collections of binary-coded ternary in the `T1B1` encoding (one trit per byte), similar to
+[PathBuf](https://doc.rust-lang.org/std/path/struct.PathBuf.html) and
+[Path](https://doc.rust-lang.org/std/path/struct.Path.html).
+
+```rust
+/// The common interface of cryptographic hash functions that follow the sponge construction and that act on ternary.
+trait Sponge {
+    /// The expected length of the input to the sponge.
+    const IN_LEN: usize;
```

Comment:
Should these always be constant?

Reply:
I assume these associated constants were added for flexibility but at the moment they are all defined to `HASH_LEN` and we surely won't have any more ternary sponge. I guess we can remove them.